### ISSUE TITLE
Web containers are not fully closed on context close in 1.4.4.RELEASE and later, preventing termination

### ISSUE THREAD

**Issue by mkw** [spring-projects/spring-boot#8224]
In 1.4.2.RELEASE and 1.4.3.RELEASE, when validation prevented startup from completing, the web server was shut down and the JVM terminated.  In 1.4.4.RELEASE and 1.5.1.RELEASE, the JVM continues running after validation failure.

Debugging traces the error to having been introduced in 0af53b3, which fixed gh-8036.  This commit introduced a `started` flag to each concrete sub-class of `org.springframework.boot.context.embedded.EmbeddedServletContainer` designed to track if the container had been started.  This flag is used to protect against multiple restarts and checked before the shutdown procedure is invoked in `stop()`.

However, it appears that each containers is partially started by `initialize()` -- e.g. in the v1.4.4.RELEASE version of `org.springframework.boot.context.embedded.tomcat.TomcatEmbeddedServletContainer#initialize`, on line 98, `this.tomcat.start()` is called.  This invocation starts the non-daemon threads that prevent shutdown.

There are two obvious ways that this might be fixed in implementations of `EmbeddedServletContainer#stop`:
1. The check for `started` could simply be removed.
2. Portions of the existing implementation that shutdown the parts of startup performed in `TomcatEmbeddedServletContainer#initialize` could be moved above the check of `started`.

I'll add a PR for the latter, as it seems safer to me.

This can be reproduced with a project consisting of the following two files:

build.gradle
```gradle
buildscript {
  ext {
    //springBootVersion = '1.4.2.RELEASE' // Running `gradle bootRun` terminates on validation failure.
    springBootVersion = '1.5.1.RELEASE' // Running `gradle bootRun` does not terminate on  validation failure.
  }
  repositories {
    mavenCentral()
  }
  dependencies {
    classpath("org.springframework.boot:spring-boot-gradle-plugin:${springBootVersion}")
  }
}

apply plugin: 'java'
apply plugin: 'org.springframework.boot'

jar {
  baseName = 'validation-stop-failure-1.5.1.RELEASE'
  version = '0.0.1-SNAPSHOT'
}

sourceCompatibility = 1.8

repositories {
  mavenCentral()
}


dependencies {
  compile('org.springframework.boot:spring-boot-starter-web')
  compile('org.springframework.boot:spring-boot-starter-validation')
}
```

src/main/java/com/coruscations/spring/boot/validationfailure/ValidationStopFailure.java
```java
package com.coruscations.spring.boot.validationfailure;

import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.boot.context.properties.ConfigurationProperties;
import org.springframework.boot.context.properties.EnableConfigurationProperties;
import org.springframework.validation.annotation.Validated;

import javax.validation.constraints.NotNull;

@SpringBootApplication
@EnableConfigurationProperties(ValidationStopFailure.NameConfiguration.class)
public class ValidationStopFailure {

  public static void main(String[] args) {
    SpringApplication.run(ValidationStopFailure.class, args);
  }

  @ConfigurationProperties(prefix = "com.coruscations.validation-stop-failure")
  @Validated
  static class NameConfiguration {

    @NotNull(message = "Name cannot be null")
    private String name;

    public String getName() {
      return name;
    }

    public void setName(String name) {
      this.name = name;
    }
  }
}
```

Assuming Gradle is in your path, (tested with 3.3), execute `gradle bootRun`.  If the Spring Boot version is set to 1.4.2.RELEASE, a build failure results.  If the Spring Boot version is set to 1.5.1.RELEASE, it hangs with the application failing to shutdown.

A thread dump reveals that the JVM is held open by this non-daemon thread:
```
"container-0" #15 prio=5 os_prio=31 tid=0x00007faba23b2800 nid=0x4f03 waiting on condition [0x00007000057fa000]
   java.lang.Thread.State: TIMED_WAITING (sleeping)
        at java.lang.Thread.sleep(Native Method)
        at org.apache.catalina.core.StandardServer.await(StandardServer.java:427)
        at org.springframework.boot.context.embedded.tomcat.TomcatEmbeddedServletContainer$1.run(TomcatEmbeddedServletContainer.java:168)
```

**Comment by philwebb** [comment]
Closing in favor of PR #8227
